daffodil-codegen-c/src/main/scala/org/apache/daffodil/codegen/c/generators/CodeGeneratorState.scala

```diff
@@ -727,22 +778,32 @@ class CodeGeneratorState(private val root: ElementBase) {
 
   // Recursively builds a hopefully unique name using the given StringBuilder
   private def buildName(sc: SchemaComponent, sb: StringBuilder): StringBuilder = {
+    // Append schema component's name
     sc match {
       case eb: ElementBase => sb ++= eb.namedQName.local += '_'
       case gd: GlobalElementDecl => sb ++= gd.namedQName.local += '_'
       case ct: GlobalComplexTypeDef => sb ++= ct.namedQName.local += '_'
       case _ => // don't include other schema components in qualified name
     }
+    // Recursively append parent schema components' names
     sc.optLexicalParent.foreach {
       buildName(_, sb)
     }
+    // Ensure name contains only legal characters for C identifiers
+    lazy val legalChars: immutable.Set[Char] =
+      Set('_') ++ ('a' to 'z') ++ ('A' to 'Z') ++ ('0' to '9')
+    for (i <- sb.indices) {
+      if (!legalChars.contains(sb.charAt(i))) {
+        sb.setCharAt(i, '_')
+      }
+    }
```

Comment:
This logic about names also exists the `cName` function. Can those be refactored into a common function so there's only one place that deals with creating legal names?

Reply:
Yes, I've refactored the duplicate code for legal C identifiers into a single `makeLegalForC` function.